File: android/src/main/java/com/regula/plugin/idv/JSONConstructor.kt
@file:Suppress("unused")

package com.regula.plugin.idv

import com.regula.idv.api.config.ApiKeyConnectionConfig
import com.regula.idv.api.config.CredentialsConnectionConfig
import com.regula.idv.api.config.PrepareWorkflowConfig
import com.regula.idv.api.config.StartWorkflowConfig
import com.regula.idv.api.config.TokenConnectionConfig
import com.regula.idv.api.models.WorkflowResult
import com.regula.idv.api.models.Workflow
import com.regula.idv.api.models.WorkflowStep
import com.regula.idv.module.BaseException
import org.json.JSONObject

fun generateCompletion(success: Any?, error: BaseException?) = mapOf(
    "success" to success,
    "error" to error?.message
).toJson()

// Config ------------------------------

fun credentialsConnectionConfigFromJSON(it: JSONObject) = CredentialsConnectionConfig(
    it.getString("baseUrl"),
    it.getString("userName"),
    it.getString("password")
).let { self ->
    self.httpTimeoutMs = it.getIntOrNull("httpTimeoutMs")
    return@let self
}

fun generateCredentialsConnectionConfig(it: CredentialsConnectionConfig) = mapOf(
    "baseUrl" to it.baseUrl,
    "userName" to it.userName,
    "password" to it.password,
    "httpTimeoutMs" to it.httpTimeoutMs,
).toJson()

fun tokenConnectionConfigFromJSON(it: JSONObject) = TokenConnectionConfig(
    it.getString("url")
)

fun generateTokenConnectionConfig(it: TokenConnectionConfig) = mapOf(
    "url" to it.baseUrl,
).toJson()

fun apiKeyConnectionConfigFromJSON(it: JSONObject) = ApiKeyConnectionConfig(
    it.getString("baseUrl"),
    it.getString("apiKey"),
    it.getIntOrNull("ttl"),
).let { self ->
    self.httpTimeoutMs = it.getIntOrNull("httpTimeoutMs")
    return@let self
}

fun generateApiKeyConnectionConfig(it: ApiKeyConnectionConfig) = mapOf(
    "baseUrl" to it.baseUrl,
    "apiKey" to it.apiKey,
    "ttl" to it.ttl,
    "httpTimeoutMs" to it.httpTimeoutMs,
).toJson()

fun prepareWorkflowConfigFromJSON(it: JSONObject) = PrepareWorkflowConfig(
    it.getString("workflowId"),
)

fun generatePrepareWorkflowConfig(it: PrepareWorkflowConfig) = mapOf(
    "workflowId" to it.workflowId,
).toJson()

fun startWorkflowConfigFromJSON(input: JSONObject?) = input?.let { json ->
    val builder = StartWorkflowConfig.Builder()

    json.getStringOrNull("locale")?.let { builder.setLocale(it) }
    json.getJSONObjectOrNull("metadata")?.let { builder.setMetadata(it) }

    builder.build()
}

fun generateStartWorkflowConfig(input: StartWorkflowConfig?) = input?.let {
    mapOf(
        "locale" to it.locale,
        "metadata" to it.metadata,
    ).toJson()
}

// Model ------------------------------

fun workflowFromJSON(input: JSONObject?) = input?.let {
    Workflow(
        it.getString("id"),
        it.getString("name"),
        it.getString("version"),
        it.getString("description"),
        it.getStringOrNull("defaultLocale"),
    )
}

fun generateWorkflow(input: Workflow?) = input?.let {
    mapOf(
        "id" to it.id,
        "name" to it.name,
        "version" to it.version,
        "description" to it.description,
        "defaultLocale" to it.defaultLocale,
    ).toJson()
}

fun workflowStepFromJSON(input: JSONObject?) = input?.let {
    WorkflowStep(
        it.getString("id"),
        it.getString("name"),
    )
}

fun generateWorkflowStep(input: WorkflowStep?) = input?.let {
    mapOf(
        "id" to it.id,
        "name" to it.name,
    ).toJson()
}

fun workflowResultFromJSON(input: JSONObject?): WorkflowResult? = input?.let {
    WorkflowResult(
        it.getString("sessionId"),
        workflowStepFromJSON(it.getJSONObjectOrNull("finalStep"))!!
    )
}

fun generateWorkflowResult(input: WorkflowResult?) = input?.let {
    mapOf(
        "sessionId" to it.sessionId,
        "finalStep" to generateWorkflowStep(it.finalStep),
    ).toJson()
}

File: ios/JSONConstructor.swift
import IDVSDK
import IDVModule

public func generateCompletion(_ value: Any?, _ error: IDVModule.BaseError?) -> [String: Any?] {
    return [
        "success": value,
        "error": error?.fullChain
    ]
}

// MARK: - Config

public func credentialsConnectionConfigFromJSON(_ data: [String: Any?]) -> CredentialsConnectionConfig {
    let result = CredentialsConnectionConfig (userName: data["userName"] as! String,
                                              password: data["password"] as! String,
                                              baseURL: data["baseUrl"] as! String)
    if let httpTimeoutMs = data["httpTimeoutMs"] as? NSNumber { result.httpTimeoutMs = httpTimeoutMs }
    return result
}

public func generateCredentialsConnectionConfig(_ data: CredentialsConnectionConfig) -> [String: Any?] {
    return [
        "baseUrl": data.baseURL,
        "userName": data.userName,
        "password": data.password,
        "httpTimeoutMs": data.httpTimeoutMs,
    ]
}

public func tokenConnectionConfigFromJSON(_ data: [String: Any?]) -> TokenConnectionConfig {
    return TokenConnectionConfig(url: data["url"] as! String)
}

public func generateTokenConnectionConfig(_ data: TokenConnectionConfig) -> [String: Any?] {
    return [
        "url": data.url
    ]
}

public func apiKeyConnectionConfigFromJSON(_ data: [String: Any?]) -> ApiKeyConnectionConfig {
    let result = ApiKeyConnectionConfig(apiKey: data["apiKey"] as! String,
                                        baseURL: data["baseUrl"] as! String,
                                        ttl: data["ttl"] as? NSNumber)
    if let httpTimeoutMs = data["httpTimeoutMs"] as? NSNumber { result.httpTimeoutMs = httpTimeoutMs }
    return result
}

public func generateApiKeyConnectionConfig(_ data: ApiKeyConnectionConfig) -> [String: Any?] {
    return [
        "baseUrl": data.baseURL,
        "apiKey": data.apiKey,
        "ttl": data.ttl,
        "httpTimeoutMs": data.httpTimeoutMs,
    ]
}

public func prepareWorkflowConfigFromJSON(_ data: [String: Any?]) -> PrepareWorkflowConfig {
    return PrepareWorkflowConfig(workflowId: data["workflowId"] as! String)
}

public func generatePrepareWorkflowConfig(_ data: PrepareWorkflowConfig) -> [String: Any?] {
    return [
        "workflowId": data.workflowId
    ]
}

public func startWorkflowConfigFromJSON(input: [String: Any?]?) -> StartWorkflowConfig {
    let result = StartWorkflowConfig.default()
    guard let data = input else { return result }
    
    if let locale = data["locale"] as? String { result.locale = locale }
    if let metadata = data["metadata"] as? [String: Any] { result.metadata = metadata }
    
    return result
}

public func generateStartWorkflowConfig(_ data: StartWorkflowConfig) -> [String: Any?] {
    return [
        "locale": data.locale,
        "metadata": data.metadata
    ]
}

// MARK: - Model

public func workflowFromJSON(_ input: [String: Any?]?) -> Workflow? {
    guard var it = input else { return nil }
    it["client"] = [String: Any]()
    it["steps"] = [WorkflowStep]()
    it["_description"] = it["description"]
    return try! Workflow(from: it.toDecoder())
}

public func generateWorkflow(_ input: Workflow?) -> [String: Any?]? {
    guard let it = input else { return nil }
    return [
        "id": it.id,
        "name": it.name,
        "version": it.version,
        "description": it._description,
        "defaultLocale": it.defaultLocale
    ]
}

public func workflowStepFromJSON(_ input: [String: Any?]?) -> WorkflowStep? {
    guard var it = input else { return nil }
    it["type"] = ""
    it["final"] = false
    it["client"] = [String: Any]()
    return try! WorkflowStep(from: it.toDecoder())
}

public func generateWorkflowStep(_ input: WorkflowStep?) -> [String: Any?]? {
    guard let it = input else { return nil }
    return [
        "id": it.id,
        "name": it.name
    ]
}

public func workflowResultFromJSON(_ input: [String: Any?]?) -> WorkflowResult? {
    guard let it = input else { return nil }
    return try! WorkflowResult(from: it.toDecoder())
}

public func generateWorkflowResult(_ input: WorkflowResult?) -> [String: Any?]? {
    guard let it = input else { return nil }
    return [
        "sessionId": it.sessionId,
        "finalStep": generateWorkflowStep(it.finalStep)
    ]
}
